crypto/chacha_drbg.c

```diff
@@ -19,44 +19,96 @@
 
 #include "chacha_drbg.h"
 
+#include <assert.h>
+#include <sha2.h>
 #include <stdint.h>
+#include <string.h>
+
+#include "chacha20poly1305/ecrypt-portable.h"
+#include "memzero.h"
 
 #define MAX(a, b) (a) > (b) ? (a) : (b)
 
-void chacha_drbg_init(CHACHA_DRBG_CTX *ctx,
-                      const uint8_t entropy[CHACHA_DRBG_SEED_LENGTH]) {
+static void derivation_function(const uint8_t *input1, size_t input1_length,
+                                const uint8_t *input2, size_t input2_length,
+                                uint8_t *output, uint32_t output_length) {
+  uint32_t blocks_number = (output_length - 1) / SHA256_DIGEST_LENGTH + 1;
+  assert(blocks_number <= 255);
+
+  SHA256_CTX ctx = {0};
+
+  for (uint8_t counter = 1; counter <= blocks_number; counter++) {
+    sha256_Init(&ctx);
+    sha256_Update(&ctx, &counter, sizeof(counter));
+    sha256_Update(&ctx, (uint8_t *)&output_length, sizeof(output_length));
+    sha256_Update(&ctx, input1, input1_length);
+    sha256_Update(&ctx, input2, input2_length);
```

Comment:
I see that the lengths are constant, so there is no reason to add them in our implementation. I'd just add a comment mentioning, that it's an approximate implementation of Hash_df and otherwise leave as is.

Reply:
Actually `derivation_function` was intended to be an exact implementation of `Hash_df`, which cannot be true for the rest of the drbg (since chacha is not a block cipher).

I fixed it so it hashes `output_length` in bits, see https://github.com/trezor/trezor-firmware/commit/37cc2f52281fd365a56b71fcac1fea3e037f37ac.

Reply:
Fixed in ca41ae51adddcf4654c7f17b0602906df8b28436.